tests/test_shell.py

```diff
@@ -199,13 +324,234 @@ def test_remove(self):
+        assert 'a' in project_a.document
+        assert 'a' in project_b.document
+        assert 'b' in project_a.document
+        assert 'b' not in project_b.document
+        with job_dst:
+            with open('test', 'r') as file:
+                assert 'x' == file.read()
+        # invalid cases
+        with pytest.raises(ExitCodeError):
+            self.call('python -m signac sync {} {} -s never -u'
+                      .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        with pytest.raises(ExitCodeError):
+            self.call('python -m signac sync {} {} -t'
+                      .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+
+    def test_sync_merge(self):
+        project_b = signac.init_project('ProjectB', os.path.join(self.tmpdir.name, 'b'))
+        self.call('python -m signac init ProjectA'.split())
+        project_a = signac.Project()
+        for i in range(4):
+            project_a.open_job({'a': i}).init()
+            project_b.open_job({'a': i}).init()
+        project_a.open_job({'c': 1}).init()
+        project_b.open_job({'b': 1}).init()
+        project_b.open_job({'a': 4}).init()
+        assert len(project_a) == 5
+        assert len(project_b) == 6
+
+        with pytest.raises(ExitCodeError):
+            self.call('python -m signac sync {} {}'
+                      .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert len(project_a) == 5
+        assert len(project_b) == 6
+
+        self.call('python -m signac sync {} {} --merge'
+                  .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert len(project_a) == 7
+        assert len(project_b) == 6
+
+    def test_sync_document(self):
+        self.call('python -m signac init ProjectA'.split())
+        project_a = signac.Project()
+        project_b = signac.init_project('ProjectB', os.path.join(self.tmpdir.name, 'b'))
+        job_src = project_a.open_job({'a': 0})
+        job_dst = project_b.open_job({'a': 0})
+
+        def reset():
+            job_src.document['a'] = 0
+            job_src.document['nested'] = dict(a=1)
+            job_dst.document['a'] = 1
+            job_dst.document['nested'] = dict(a=2)
+
+        reset()
+        assert job_dst.document != job_src.document
+        with pytest.raises(ExitCodeError):
+            self.call('python -m signac sync {} {}'
+                      .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+
+        self.call('python -m signac sync {} {} --no-key'
+                  .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+        assert job_dst.document['a'] != job_src.document['a']
+        assert job_dst.document['nested'] != job_src.document['nested']
+
+        reset()     # only sync a
```

Comment:
Split these comments onto their own lines, above `reset()`. Also add a comment explaining what is happening for all the options (`--no-key`, `--all-key`, etc.).